Source/WebCore/rendering/style/StyleCustomPropertyData.h

```diff
@@ -35,36 +36,111 @@ class StyleCustomPropertyData : public RefCounted<StyleCustomPropertyData> {
 public:
     static Ref<StyleCustomPropertyData> create() { return adoptRef(*new StyleCustomPropertyData); }
     Ref<StyleCustomPropertyData> copy() const { return adoptRef(*new StyleCustomPropertyData(*this)); }
-    
-    bool operator==(const StyleCustomPropertyData& other) const
-    {
-        if (values.size() != other.values.size())
-            return false;
 
-        for (auto& entry : values) {
-            auto otherEntry = other.values.find(entry.key);
-            if (otherEntry == other.values.end() || !entry.value->equals(*otherEntry->value))
-                return false;
-        }
+    bool operator==(const StyleCustomPropertyData& other) const;
+    bool operator!=(const StyleCustomPropertyData& other) const { return !(*this == other); }
 
-        return true;
-    }
+    const CSSCustomPropertyValue* get(const AtomString&) const;
+    void set(const AtomString&, Ref<const CSSCustomPropertyValue>&&);
 
-    bool operator!=(const StyleCustomPropertyData& other) const { return !(*this == other); }
-    
-    void setCustomPropertyValue(const AtomString& name, Ref<const CSSCustomPropertyValue>&& value)
-    {
-        values.set(name, WTFMove(value));
-    }
+    unsigned size() const { return m_sizeExcludingInheritedValues + (m_inheritedValues ? m_inheritedValues->m_sizeExcludingInheritedValues : 0); }
 
-    CustomPropertyValueMap values;
+    void forEach(const Function<void(const KeyValuePair<AtomString, RefPtr<const CSSCustomPropertyValue>>&)>&) const;
+    AtomString keyAtIndex(unsigned) const;
 
 private:
     StyleCustomPropertyData() = default;
-    StyleCustomPropertyData(const StyleCustomPropertyData& other)
-        : RefCounted<StyleCustomPropertyData>()
-        , values(other.values)
-    { }
+    StyleCustomPropertyData(const StyleCustomPropertyData&);
+
+    RefPtr<const StyleCustomPropertyData> m_inheritedValues;
+    CustomPropertyValueMap m_values;
+    unsigned m_sizeExcludingInheritedValues;
 };
 
+inline StyleCustomPropertyData::StyleCustomPropertyData(const StyleCustomPropertyData& other)
+    : RefCounted<StyleCustomPropertyData>()
+    , m_inheritedValues(!other.m_inheritedValues && !other.m_values.isEmpty() ? &other : other.m_inheritedValues.get())
+    , m_values(!other.m_inheritedValues && !other.m_values.isEmpty() ? CustomPropertyValueMap() : other.m_values)
+{
+}
+
+inline const CSSCustomPropertyValue* StyleCustomPropertyData::get(const AtomString& name) const
+{
+    if (auto* value = m_values.get(name))
+        return value;
+    if (m_inheritedValues) {
+        ASSERT(!m_inheritedValues->m_inheritedValues);
+        if (auto* value = m_inheritedValues->m_values.get(name))
+            return value;
+    }
+    return nullptr;
+}
+
+inline void StyleCustomPropertyData::set(const AtomString& name, Ref<const CSSCustomPropertyValue>&& value)
```

Comment:
This relies in properties never being set after style construction (or you'd mutate the inheriting styles too). Fortunately this should be the case.

Reply:
I think the use of `DataRef`s on `StyleRareNonInheritedData` and `StyleRareInheritedData` should prevent this even if we did set property values after style construction. I could change the `RefPtr` here to be a `DataRef` to maybe make that clearer, even though we don't modify property values through that `RefPtr`.

Reply:
Oh, I guess `DataRef` doesn't allow a null pointer. I'll leave it as `RefPtr`.